packages/core/src/database/index.ts

```diff
@@ -0,0 +1,1211 @@
+                msg: `Reverting operations prior to finalized checkpoint (timestamp=${blockTimestamp} chainId=${chainId} block=${blockNumber})`,
+              });
+
+              for (const tableName of Object.keys(getTables(args.schema))) {
+                await revert({
+                  tableName,
+                  checkpoint: previousApp.checkpoint,
+                  tx,
+                });
+              }
+
+              return {
+                status: "success",
+                checkpoint: previousApp.checkpoint,
+              } as const;
+            }
+
+            /**
+             * If table name collision, exit
+             * Else, start
+             *
+             * At this point in the control flow, the previous app has a
+             * different build ID or a zero "checkpoint". We need to drop the
+             * previous app's tables and create new ones.
+             */
+
+            await tx
+              .updateTable("_ponder_meta")
+              .set({ value: encodeApp(newApp) })
+              .where("key", "=", "app")
+              .execute();
+
+            args.common.logger.debug({
+              service: "database",
+              msg: `Acquired lock on schema '${namespace}' previously used by build '${previousApp.build_id}'`,
+            });
+
+            // Drop old tables
+
+            for (const tableName of previousApp.table_names) {
+              await tx.schema
+                .dropTable(`_ponder_reorg_${tableName}`)
+                .ifExists()
+                .execute();
+
+              await tx.schema.dropTable(tableName).ifExists().execute();
+
+              args.common.logger.debug({
+                service: "database",
+                msg: `Dropped '${tableName}' table left by previous build`,
+              });
+            }
+
+            await createUserTables();
+            await createReorgTables();
+
+            return {
+              status: "success",
+              checkpoint: encodeCheckpoint(zeroCheckpoint),
+            } as const;
+          }),
+        );
+
+      let result = await attempt();
+      if (result.status === "locked") {
+        const duration = result.expiry - Date.now();
+        args.common.logger.warn({
+          service: "database",
+          msg: `Schema '${namespace}' is locked by a different Ponder app`,
+        });
+        args.common.logger.warn({
+          service: "database",
+          msg: `Waiting ${formatEta(duration)} for lock on schema '${namespace} to expire...`,
+        });
+
+        await wait(duration);
+
+        result = await attempt();
+        if (result.status === "locked") {
+          throw new NonRetryableError(
+            `Failed to acquire lock on schema '${namespace}'. A different Ponder app is actively using this database.`,
+          );
+        }
+      }
+
+      heartbeatInterval = setInterval(async () => {
+        try {
+          const app = await getApp(orm.internal);
```

Comment:
Seems a bit weird that we need to get the app each time. Could instead use JSON operators (https://www.sqlite.org/json1.html#jset) to only update the heartbeat_at property.

Something like this:
```
.set({
  value: ksql`json_set(value, '$.heartbeat_at', ${encodeApp({ ...app!, heartbeat_at: Date.now() })})`,
})
```